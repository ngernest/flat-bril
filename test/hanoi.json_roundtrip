{
  "functions": [
    {
      "args": [
        {
          "name": "disks",
          "type": "int"
        },
        {
          "name": "src",
          "type": "int"
        },
        {
          "name": "dst",
          "type": "int"
        },
        {
          "name": "spare",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "zero",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "disks",
            "zero"
          ],
          "dest": "pos",
          "funcs": [],
          "labels": [],
          "op": "gt",
          "type": "bool"
        },
        {
          "args": [
            "pos"
          ],
          "funcs": [],
          "labels": [
            "then",
            "else"
          ],
          "op": "br"
        },
        {
          "label": "then"
        },
        {
          "dest": "one",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "disks",
            "one"
          ],
          "dest": "above",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "above",
            "src",
            "spare",
            "dst"
          ],
          "funcs": [
            "hanoi"
          ],
          "labels": [],
          "op": "call"
        },
        {
          "args": [
            "src",
            "dst"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "args": [
            "above",
            "spare",
            "dst",
            "src"
          ],
          "funcs": [
            "hanoi"
          ],
          "labels": [],
          "op": "call"
        },
        {
          "label": "else"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "hanoi"
    },
    {
      "args": [
        {
          "name": "disks",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "src",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "dest": "dst",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "dest": "spare",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "disks",
            "src",
            "dst",
            "spare"
          ],
          "funcs": [
            "hanoi"
          ],
          "labels": [],
          "op": "call"
        }
      ],
      "name": "main"
    }
  ]
}
