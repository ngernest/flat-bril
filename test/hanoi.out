@hanoi
	op: const		dest: zero		type: int		value: 0	
	op: gt   		dest: pos		type: bool		args: ["disks", "zero"]	
	op: br   		args: ["pos"]		labels: ["then", "else"]
.then
	op: const		dest: one		type: int		value: 1	
	op: sub  		dest: above		type: int		args: ["disks", "one"]	
	op: call 		args: ["above", "src", "spare", "dst"]		funcs: ["hanoi"]
	op: print		args: ["src", "dst"]	
	op: call 		args: ["above", "spare", "dst", "src"]		funcs: ["hanoi"]
.else
	op: ret  	
@main
	op: const		dest: src		type: int		value: 0	
	op: const		dest: dst		type: int		value: 2	
	op: const		dest: spare		type: int		value: 1	
	op: call 		args: ["disks", "src", "dst", "spare"]		funcs: ["hanoi"]
